packages/firestore/src/core/firestore_client.ts

```diff
@@ -404,7 +404,7 @@ export class FirestoreClient {
     if (this.clientTerminated) {
       return;
     }
-    this.asyncQueue.enqueueAndForget(() => {
+    this.asyncQueue.enqueueRetryable(() => {
```

Comment:
Why is this necessary? I thought that if we failed we tore down the streams? At that point haven't we effectively unlistened anyway? All that's missing is updating our in-memory bookkeeping to indicate that we don't want to restart the target when IndexedDB becomes available again.

One other concern is that if IndexedDB is failing we've likely been put in the background and may not come back for a while--possibly more than the 30 minutes for which a resume token means free query resumption. If we were asleep that long the user would be charged for the query anew--for a query they were trying to stop listening to. If all we're losing is the resume token and the sequence numbers for exact GC ordering, it seems better to let the unlisten succeed despite failing to write. 

So in the primary tab, if we fail trying to save metadata data about unlistening that's not really a big deal. Can a secondary tab fail trying to tell the primary that it wants to unlisten? It's that case we should probably retry.

If this squares with your understanding, it seems like this line may be accurate: we do want to retry unlistens but only when it's the secondary tab asking the primary tab to do it. Primary tabs should just succeed.

It seems like unlisten should succeed regardless of the indexeddb failure.

Reply:
This code path calls into LocalStore to invoke "Release Target" and then calls RemoteStore. We could change the plumbing here to drive this from RemoteStore, in which case RemoteStore could go offline and do its own retry. That sounds a like a fair bit of code churn that may lead to some unexpected side effects - the mapping between the `lastListen` state in EventManager and the active list of targets in SyncEngine/LocalStore may go out of sync.

One easy thing we could do here is just swallow IndexedDB failures if "Release Target" fails, which would even remove some of the newly added error handling in RemoteStore. Do you think that's worth exploring?

Reply:
> just swallow IndexedDB failures if "Release Target" fails

This is what I was suggesting: releasing a target in the LocalStore is just about making sure the final updates are made for GC metadata (and the in-memory tracking of the target). It may be that even LocalStore itself may need to change so that even if the transaction fails it will remove the target from its own set of live things.

I'm not sure this undoes the error handling in RemoteStore, since it's really only unsubscribe can work this way.

Reply:
I updated the PR to ignore errors in "Release target". Unfortunately, since a failed Target can generate a "fake" Remote Event (for Limbo resolutions), the code path that calls into this from RemoteStore still has to use the IndexedDB fallback.